[5.8][WIP] Support eager loading with limit

staudenmeir opened the pull request:
There is currently no way to limit eager loading results *per parent*:

```php
class User extends Model {
    public function posts() {
        return $this->hasMany(Post::class)->latest()->limit(5);
    }
}

User::with('posts')->get(); 
```

```sql
# Loads 5 posts for all users combined.
select * from "posts" where "posts"."user_id" in (1, 2, 3) order by "created_at" desc limit 5
```

There are multiple issues about this (#16217, #4835, #18014) and it also comes up regularly on StackOverflow etc.

We can support this feature with a [window function](https://en.wikipedia.org/wiki/SQL_window_function) (we are already using this technique to implement `OFFSET` queries on SQL Server):

```sql
select * from (
  select *, row_number() over (partition by "user_id" order by "created_at" desc) as row_num
  from "posts"
  where "posts"."user_id" in (1, 2, 3)
) as temp_table
where row_num <= 5
order by row_num
```

The query is supported by all four databases, but requires very recent versions of MySQL and SQLite:

 - MySQL 8.0 (2018-07-27): We could support older versions with [this](https://softonsofa.com/tweaking-eloquent-relations-how-to-get-n-related-models-per-parent/) workaround.
 - PostgreSQL 9.3 (2013)
 - SQLite 3.25 (2018-09-15)
 - SQL Server 2008

I wrote a sample implementation for PostgreSQL because it's the only supported database in Homestead at the moment.

The relationships override the query builder's `limit()` and call the new `partitionLimit()` method. By checking the parent model's `exists` property, we can detect eager loading. This is necessary to support lazy loading (`$user->posts`). The window function would also work for lazy loading but the database might not support it.

Before I implement the other databases: Would you consider this for the core?

staudenmeir committed:
Support eager loading with limit

deleugpn commented:
Promising.

X-Coder264 commented:
This looks very nice. It'd be great if this would be in the core.

Lelectrolux commented:
Currently using my own version of the mysql workaround, happy to see that mysql 8.0 helps to clean the madness even if I'm on my own, but I would love that in laravel's core.

For example of use in real life, it's a needs that comes often when doing travelling best offers (like find the 3 best price for this plane/train in the next week). My current work is full of that.

staudenmeir commented:
I added the remaining implementations. We can use the same SQL in all four databases. The only exception is that SQL Server always requires an `ORDER BY` clause (we use a dummy clause if necessary).

- MySQL:
I implemented the mentioned workaround for MySQL < 8.0. It works for `HasMany` relationships but not for `BelongsToMany` or `HasManyThrough`. For some reason, MySQL ignores the `ORDER BY` clause and always uses the row insertion order ([fiddle](https://www.db-fiddle.com/f/cFq2sh6oEVVRudJp2qv7Zb/0)). Does anyone have an idea on how to fix this?

- SQLite:
SQLite 3.25.0 has only been released a month ago, so this feature doesn't work with the current PHP distributions. You have to compile the latest `pdo_sqlite` extension yourself (I tested it on Windows). So, no integration tests.
Since SQLite is mainly used for testing, I suggest that we completely ignore the partition limit on unsupported versions. This PR is "only" a performance improvement and I don't think it's worth breaking the users' tests.

- SQL Server:
I moved `compileOver()` and `compileTableExpression()` up to `Grammar` and refactored them to work with the new features.

miki131 commented:
@staudenmeir what about some `LimitTrait` and move there `take` and `limit` functions?

staudenmeir commented:
That's not possible for `limit()`: All three methods are different from each other.

miki131 commented:
@staudenmeir take a look on `getExistenceCompareKey()` method. It can be useful for `BelongsToMany.php` and `HasOneOrMany.php`. Later i'll try find out about third relation.

staudenmeir commented:
Yes, but we still need a separate implementation for `HasManyThrough`. I don't think it makes sense to create a trait for two relationships and override it in the third.

miki131 commented:
@staudenmeir You can create `getExistenceCompareKey` method for `HasManyThrough`. Also override `getParent` method. I think it can be useful in future and obviously help you separate code into trait.

staudenmeir commented:
@miki131 The `HasManyThrough` relationship does have a `$parent` property, so `getParent()` can't just return the `$farParent`.

ludo237 commented:
What about MariaDB? 

Promising feature tho

staudenmeir commented:
@ludo237 I'm working on that. MariaDB has been supporting window functions since 10.2, but they apparently don't work in `ONLY_FULL_GROUP_BY` mode. Looks like a bug.

taylorotwell commented:
Do any other frameworks support this?

staudenmeir commented:
[Rails](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html) doesn't support it:

> If you eager load an association with a specified :limit option, it will be ignored, returning all the associated objects.

I don't know about Symfony, I couldn't find anything. So probably not.

taylorotwell commented:
Honestly I don't see a big incentive to take on the maintenance burden of maintaining this functionality. If it's not even supported by Rails that indicates to me it's not that widely useful.

staudenmeir commented:
I released the code as a package: https://github.com/staudenmeir/eloquent-eager-limit

yaquawa commented:
Rejected as usual. 

Tips: To get this into core, the best way is send a PR to Rails.

atrauzzi commented:
Just hit this today, a bit confused as to why it was rejected.  This would be a great - if not essential - addition to Eloquent.  If people have the syntax available, it stands to reason they'll assume this is possible.

driesvints commented:
Hey @atrauzzi, there's a package for this which you may use: https://github.com/staudenmeir/eloquent-eager-limit

tillkruss commented:
When using PostgreSQL use `DISTINCT ON`: https://github.com/laravel/framework/issues/4835#issuecomment-526341623

staudenmeir commented:
Eager loading with limit will be supported natively in Laravel 11 (#49695).
